src/github.com/matrix-org/dendrite/clientapi/routing/login.go

```diff
@@ -50,25 +64,112 @@ type loginResponse struct {
 	DeviceID    string                       `json:"device_id"`
 }
 
-func passwordLogin() loginFlows {
+func defaultPasswordLogin() loginFlows {
 	f := loginFlows{}
-	s := flow{"m.login.password", []string{"m.login.password"}}
+	s := flow{string(PasswordBased), []string{string(PasswordBased)}}
 	f.Flows = append(f.Flows, s)
 	return f
 }
 
+func handlePasswordLogin(
+	r loginRequest, accountDB *accounts.Database, deviceDB *devices.Database,
+	req *http.Request, cfg config.Dendrite) *util.JSONResponse {
+
+	localpart := r.User
+	acc, err := accountDB.GetAccountByPassword(req.Context(), localpart, r.Password)
+
+	if err != nil {
+		// Technically we could tell them if the user does not exist by checking if err == sql.ErrNoRows
+		// but that would leak the existence of the user.
+		return &util.JSONResponse{
+			Code: 403,
+			JSON: jsonerror.Forbidden("username or password was incorrect, or the account does not exist"),
+		}
+	}
+
+	token, err := auth.GenerateAccessToken()
+	if err != nil {
+		httputil.LogThenError(req, err)
+	}
+
+	// TODO: Use the device ID in the request
+	dev, err := deviceDB.CreateDevice(
+		req.Context(), acc.Localpart, nil, token, r.InitialDisplayName,
+	)
+	if err != nil {
+		return &util.JSONResponse{
+			Code: 500,
+			JSON: jsonerror.Unknown("failed to create device: " + err.Error()),
+		}
+	}
+
+	return &util.JSONResponse{
+		Code: 200,
+		JSON: loginResponse{
+			UserID:      dev.UserID,
+			AccessToken: dev.AccessToken,
+			HomeServer:  cfg.Matrix.ServerName,
+			DeviceID:    dev.ID,
+		},
+	}
+}
+
+func handleTokenLogin(
+	r loginRequest, deviceDB *devices.Database,
+	req *http.Request, cfg config.Dendrite) *util.JSONResponse {
+	if r.Token == "" {
+		return &util.JSONResponse{
+			Code: 401,
+			JSON: jsonerror.MissingToken("missing access token"),
+		}
+	}
+
+	dev, err := deviceDB.GetDeviceByAccessToken(req.Context(), r.Token)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return &util.JSONResponse{
+				Code: 401,
+				JSON: jsonerror.UnknownToken("Unknown token"),
+			}
+		}
+		return &util.JSONResponse{
+			Code: 401,
+			JSON: jsonerror.Unknown("Unexpected Server error occurred"),
+		}
+	}
+
+	if dev.ID != r.DeviceID {
+		// The access token specified in the request was generated for a
+		// different device.
+		return &util.JSONResponse{
+			Code: 403,
+			JSON: jsonerror.Forbidden("The access token was generated for a different device."),
```

Comment:
Remove the "." from the end of the error.

Also not sure if we should tell this information or not. It's certainly useful, but could it be used in any malicious way?

Reply:
"Access token not valid for this device" would be less revealing ?